src/test/java/cloud/eppo/UtilsTest.java

```diff
@@ -25,6 +28,42 @@ public void testGetMd5Hash() {
     assertEquals("429fb7196ccb2978443a0de8da180e00", getMD5Hex("input-34"));
   }
 
+  @Test
+  public void testGetMd5HashThreadSafe() {
+    final AtomicBoolean interferenceEncountered = new AtomicBoolean(false);
+    int numThreads = 2;
+    ExecutorService pool = Executors.newFixedThreadPool(2);
+    try {
+      for (int i = 0; i < numThreads; i += 1) {
+        pool.execute(
+            () -> {
+              if (testForMd5Interference()) {
+                interferenceEncountered.set(true);
+              }
+            });
+      }
+      pool.shutdown();
+    } finally {
+      try {
+        assertTrue(pool.awaitTermination(5, TimeUnit.SECONDS));
+      } catch (InterruptedException ex) {
+        fail();
+      }
+      assertFalse(interferenceEncountered.get());
+    }
+  }
+
+  private boolean testForMd5Interference() {
+    boolean interferenceEncountered = false;
+    for (int i = 0; i < 100; i += 1) {
+      if (!getMD5Hex("input-62").equals("448614887a99f16179b400cfccceb72d")) {
```

Comment:
👍 In other words, in every thread, the hash must be the same. Interference would mean appending to the hashed input -> different hash outputs

Reply:
Exactly! If two threads add to the digest at the same time between `reset()` calls they both get bad hashes.